scipy/optimize/_optimize.py

```diff
@@ -1404,25 +1409,31 @@ def _minimize_bfgs(fun, x0, args=(), jac=None, callback=None,
         if (gnorm <= gtol):
             break
 
+        #  See Chapter 5 in  P.E. Frandsen, K. Jonasson, H.B. Nielsen,
+        #  O. Tingleff: "Unconstrained Optimization", IMM, DTU.  1999.
+        #  These notes are available here:
+        #   http://www2.imm.dtu.dk/documents/ftp/publlec.html
+        if (alpha_k*vecnorm(pk) <= xtol*(xtol + vecnorm(xk))):
+            break
+
         if not np.isfinite(old_fval):
             # We correctly found +-Inf as optimal value, or something went
             # wrong.
             warnflag = 2
             break
 
-        rhok_inv = np.dot(yk, sk)
-        # this was handled in numeric, let it remaines for more safety
-        if rhok_inv == 0.:
-            rhok = 1000.0
-            if disp:
-                print("Divide-by-zero encountered: rhok assumed large")
-        else:
-            rhok = 1. / rhok_inv
-
-        A1 = I - sk[:, np.newaxis] * yk[np.newaxis, :] * rhok
-        A2 = I - yk[:, np.newaxis] * sk[np.newaxis, :] * rhok
-        Hk = np.dot(A1, np.dot(Hk, A2)) + (rhok * sk[:, np.newaxis] *
-                                                 sk[np.newaxis, :])
+        # The inner product dot(yk,sk) must be strictly positive otherwise
```

Comment:
If you don't hear back from anyone in a week, I'd just undo this part. There was some hesitation to add it originally. https://github.com/scipy/scipy/issues/1261#issuecomment-17022323.

If we do add it, I wonder if `_epsilon` should be `epsilon` so it can be adjusted.

Reply:
got it, I'll make that change in a week if so

Reply:
Please adjust the default value and docs as suggested above while you're at it.

Reply:
Just committed those, was planning to change this part separately in a week in a separate commit!